### PR TITLE
brew tests: fix for ARM

### PR DESCRIPTION
**Pull request by xu-cheng**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/master/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/master/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----

### CODE REVIEW

File: Library/Homebrew/test/formula_info_spec.rb [hunk 1]
```diff
@@ -5,6 +5,8 @@
 
 describe FormulaInfo, :integration_test do
   it "tests the FormulaInfo class" do
+    ENV["HOMEBREW_DEVELOPER"] = "1"
```
Comment (MikeMcQuaid) [inline Library/Homebrew/test/formula_info_spec.rb]: Why is this change needed?

File: Library/Homebrew/test/support/helper/spec/shared_context/integration_test.rb [hunk 2]
```diff
@@ -78,7 +78,8 @@ def brew(*args)
       "HOMEBREW_BREW_FILE"        => HOMEBREW_PREFIX/"bin/brew",
       "HOMEBREW_INTEGRATION_TEST" => command_id_from_args(args),
       "HOMEBREW_TEST_TMPDIR"      => TEST_TMPDIR,
-      "HOMEBREW_DEVELOPER"        => ENV["HOMEBREW_DEVELOPER"],
+      "HOMEBREW_DEVELOPER"        => "1",
```
Comment (MikeMcQuaid) [inline Library/Homebrew/test/support/helper/spec/shared_context/integration_test.rb]: Why is this change needed?

File: Library/Homebrew/test/support/helper/spec/shared_context/integration_test.rb [hunk 3]
```diff
@@ -78,7 +78,8 @@ def brew(*args)
...
+      "HOMEBREW_RUBY_PATH"        => ENV["HOMEBREW_RUBY_PATH"],
```
Comment (MikeMcQuaid) [inline Library/Homebrew/test/support/helper/spec/shared_context/integration_test.rb]: Why is this change needed? I'd be intrigued if this has any impact on https://github.com/Homebrew/brew/issues/6579.

File: Library/Homebrew/test/formula_info_spec.rb (hunk 1 above)
Comment (xu-cheng) [reply Library/Homebrew/test/formula_info_spec.rb]: These changes are needed because the following tests will call `brew` in sub shell.

In my test with portable ruby, the correct ruby can only be picked up if both `HOMEBREW_DEVELOPER` and `HOMEBREW_RUBY_PATH` are set.

So I manually set `HOMEBREW_DEVELOPER` here.

File: Library/Homebrew/test/support/helper/spec/shared_context/integration_test.rb (hunk 3 above)
Comment (xu-cheng) [reply Library/Homebrew/test/support/helper/spec/shared_context/integration_test.rb]: > I'd be intrigued if this has any impact on #6579.

I am not sure, but I don't think they are related.

File: Library/Homebrew/test/support/helper/spec/shared_context/integration_test.rb (hunk 2 above)
Comment (xu-cheng) [reply Library/Homebrew/test/support/helper/spec/shared_context/integration_test.rb]: Similar as above, they are needed to ensure sub brew process to pick up correct ruby.

File: Library/Homebrew/test/formula_info_spec.rb (hunk 1 above)
Comment (MikeMcQuaid) [reply Library/Homebrew/test/formula_info_spec.rb]: https://github.com/Homebrew/brew/blob/master/Library/Homebrew/utils/ruby.sh#L14-L17 is what unsets `HOMEBREW_RUBY_PATH` if `HOMEBREW_DEVELOPER` isn't set. As a result, `HOMEBREW_DEVELOPER` should only be needed in cases where `/usr/local/bin/brew` is called rather than `HOMEBREW_LIBRARY_PATH/"brew.rb"` (as with the integration tests). If it's only needed in this single integration test then the `FormulaInfo` class should be modified to run the same way as the integration tests are being run.

Comment (xu-cheng) [reply Library/Homebrew/test/formula_info_spec.rb]: The issue is that `HOMEBREW_BREW_FILE` is called as a sub process in normal brew code (outside test case)
 
For example: 
https://github.com/Homebrew/brew/blob/1798deda057c409cc13b9e85a6f88419713db0a3/Library/Homebrew/formula_info.rb#L16

Comment (MikeMcQuaid) [reply Library/Homebrew/test/formula_info_spec.rb]: Yup, I understand and agree that should be probably changed to instead use `brew.rb` instead. To be explicit, though: this isn't a fix for ARM, it's a fix for running `brew tests` with a custom `HOMEBREW_RUBY_PATH`.

This doesn't explain why the other changes are needed.

Comment (xu-cheng) [reply Library/Homebrew/test/formula_info_spec.rb]: > though: this isn't a fix for ARM, it's a fix for running brew tests with a custom HOMEBREW_RUBY_PATH.

Yes. I just put this fix with other fixes in the same PR.

> This doesn't explain why the other changes are needed.

I have removed the changes on `integration_test.rb`. They should not be necessary.

Comment (MikeMcQuaid) [reply Library/Homebrew/test/formula_info_spec.rb]: > Yes. I just put this fix with other fixes in the same PR.

Ok. Could you try to fill out the PR description in future? It makes reviewing easier.

Comment (MikeMcQuaid) [reply Library/Homebrew/test/formula_info_spec.rb]: I'd like to avoid setting `HOMEBREW_DEVELOPER` here. Either refactor `formula_info` to call with `brew.rb` (as integration tests do) or make this assignment conditional on a custom Ruby path somehow.

Comment (xu-cheng) [reply Library/Homebrew/test/formula_info_spec.rb]: > I'd like to avoid setting HOMEBREW_DEVELOPER here.

Why? This only affects one test case. 

>   Either refactor formula_info to call with brew.rb (as integration tests do) 

I don't think the following code in `formula_info` is easy to read comparing with current one. Noted that they are not test code.
```ruby
json = Utils.popen_read(
      RUBY_PATH,
      "-W0",
      "-I", $LOAD_PATH.join(File::PATH_SEPARATOR),
      HOMEBREW_LIBRARY_PATH/"brew.rb",
      "info",
      "--json=v1",
      name
    )
```

> make this assignment conditional on a custom Ruby path somehow.

There is no way to tell different between a custom ruby path here and normal ruby from `PATH`. Noted that in Linux, using Ruby from system is normal behaviour. Further, it makes no sense to make test behaviour different because different ruby are used.

Comment (xu-cheng) [reply Library/Homebrew/test/formula_info_spec.rb]: If you really want to make code more clean, we should change https://github.com/Homebrew/brew/blob/master/Library/Homebrew/utils/ruby.sh#L14-L17 to use a different flag for testing purpose. 

For example,
```bash
  if [[ -z "$HOMEBREW_TEST_RUBY_PATH" ]]
  then
    unset HOMEBREW_RUBY_PATH
  fi
```

Comment (MikeMcQuaid) [reply Library/Homebrew/test/formula_info_spec.rb]: > Why? This only affects one test case.

It's a hack that's only necessary with a custom `HOMEBREW_RUBY_PATH` and `HOMEBREW_DEVELOPER` often changes the behaviour of commands. A custom `HOMEBREW_RUBY_PATH` is not something we support for non-developers.

> I don't think the following code in `formula_info` is easy to read comparing with current one. Noted that they are not test code.

It's not as easy but it is more correct. We could provide a helper method if abstraction would be better.

Comment (xu-cheng) [reply Library/Homebrew/test/formula_info_spec.rb]: I think my approach in https://github.com/Homebrew/brew/pull/6617#discussion_r336498383 should avoid problems you mentioned while still keeping the code easy to read.

Comment (MikeMcQuaid) [reply Library/Homebrew/test/formula_info_spec.rb]: That's adding another undocumented conditional for an edge-case only needed for running tests in an unsupported configuration. I'm sorry, I disagree.

Comment (xu-cheng) [reply Library/Homebrew/test/formula_info_spec.rb]: OK, updated PR as you suggest.
